python/rikai/parquet/writer.py

```diff
@@ -0,0 +1,125 @@
+        resulting rikai dataset readable by spark
+    """
+    schema_json_str = schema.json()
+    converted_df = conv(df, json.loads(schema_json_str))  # convert Rikai types
+    # The spark schema is required so Spark can read it successfully
+    table = pa.Table.from_pandas(converted_df).replace_schema_metadata(
+        {RikaiDataset.SPARK_PARQUET_ROW_METADATA: schema_json_str}
+    )
+    pq.write_to_dataset(table, dest_path, use_legacy_dataset=False)
+
+
+def conv(df: pd.DataFrame, schema: dict):
+    """Convert the given pandas dataframe to native types
+
+    Returns
+    -------
+    converted: pd.DataFrame
+        A dataframe where all Rikai types have been converted
+    """
+    assert schema["type"] == "struct"
+    converted = {}
+    for field in schema["fields"]:
+        name = field["name"]
+        ser = df[name]
+        typ = field["type"]
+        if not isinstance(typ, dict):
+            converted[name] = ser
+        elif is_udt_type(typ):
+            converted[name] = ser.apply(lambda x: _conv_udt(x, typ))
+        elif typ["type"] == "array":
+            converted[name] = ser.apply(
+                lambda arr: [conv_elm(x, typ["elementType"]) for x in arr]
+            )
+        elif typ["type"] == "struct":
+            converted[name] = ser.apply(lambda d: conv_elm(d, typ))
+        else:
+            converted[name] = ser
+    return pd.DataFrame(converted)
+
+
+def is_udt_type(typ: dict):
+    """Check whether a given field type information is a udt"""
+    if typ["type"] == "udt":
+        udt = RikaiDataset._find_udt(typ["pyClass"])
+        return udt is not None
+    return False
+
+
+def conv_elm(elm, schema):
+    """Convert a single value given it's schema information"""
+    if is_udt_type(schema):
+        return elm.to_dict()
+    elif schema["type"] == "array":
+        return [conv_elm(x, schema["elementType"]) for x in elm]
+    elif schema["type"] == "struct":
+        converted = {}
+        for field in schema["fields"]:
+            name = field["name"]
+            typ = field["type"]
+            if not isinstance(typ, dict):
+                converted[name] = elm[name]
+            elif is_udt_type(typ):
+                converted[name] = _conv_udt(elm[name], typ)
+            elif typ["type"] == "array":
+                converted[name] = [
+                    conv_elm(x, typ["elementType"]) for x in elm[name]
+                ]
+            elif typ["type"] == "struct":
+                converted[name] = conv_elm(elm[name], typ)
+            else:
+                converted[name] = elm[name]
+        return converted
+    else:
+        return elm
+
+
+def _conv_udt(v, typ):
+    """Convert the UDT value to a serializable format"""
+    if isinstance(v, ToDict):
+        return v.to_dict()
+    else:
+        assert typ["type"] == "udt"
+        udt = RikaiDataset._find_udt(typ["pyClass"])
+        return udt.serialize(v)
```

Comment:
we can probably to make a new interface to make the rikai UDT type directly returns a dict, instead of using spark's UDT interface

Reply:
> So this one still needs pyspark dependency (returns `pyspark.sql.Row`, just not live spark)?

yeah, and also the schema

Reply:
> we can probably to make a new interface to make the rikai UDT type directly returns a dict, instead of using spark's UDT interface

yeah right now what we have isn't ideal. i'm using the ToDict mixin for now.